### PR TITLE
[TRA-461] define `MsgWithdrawFromVault` proto

### PR DESCRIPTION
**Pull request by ttl33**
### Changelist
1. defined `MsgWithdrawFromVault` proto
2. added an empty `msg_server` impl
3. registered msg interface + msg categorization
4. ran `make proto-gen` and `cd indexer/packages/v4-protos && npm run build:proto`

### Test Plan
n/a

### Author/Reviewer Checklist
- [ ] If this PR has changes that result in a different app state given the same prior state and transaction list, manually add the `state-breaking` label.
- [ ] If the PR has breaking postgres changes to the indexer add the `indexer-postgres-breaking` label.
- [ ] If this PR isn't state-breaking but has changes that modify behavior in `PrepareProposal` or `ProcessProposal`, manually add the label `proposal-breaking`.
- [ ] If this PR is one of many that implement a specific feature, manually label them all `feature:[feature-name]`.
- [ ] If you wish to for mergify-bot to automatically create a PR to backport your change to a release branch, manually add the label `backport/[branch-name]`.
- [ ] Manually add any of the following labels: `refactor`, `chore`, `bug`.

## Summary by CodeRabbit

- **New Features**
  - Introduced functionality for withdrawing assets from a vault to a subaccount.
  - Added a new interface and supporting methods for `MsgWithdrawFromVault` and `MsgWithdrawFromVaultResponse`.

- **Documentation**
  - Updated descriptions for deposit and withdrawal messages to clarify their actions.

- **Chores**
  - Enhanced the interface registry configuration to include new withdrawal messages.

### CODE REVIEW

File: proto/dydxprotocol/vault/tx.proto [hunk 1]
```diff
@@ -39,6 +39,21 @@ message MsgDepositToVault {
...
+// MsgWithdrawFromVault withdraws the specified amount of from the vault to the subaccount.
+message MsgWithdrawFromVault  {
+	// The vault to withdraw from.
+	VaultId vault_id = 1;
+
+	// The subaccount to withdraw to.
+	dydxprotocol.subaccounts.SubaccountId subaccount_id = 2;
+
+	// Number of shares to withdraw.
+	uint64 shares = 3;
```
Comment (ttl33) [inline proto/dydxprotocol/vault/tx.proto:51]: the [original spec](https://www.notion.so/dydx/x-vault-caf6cd94c80e448a96a70fe20051aa1c?pvs=4#95a1b5dc21104b18a9b2aff718a32b6e) calls for `quote_quantums` but i think it makes more sense to specify the # of shares. wdyt?

File: proto/dydxprotocol/vault/tx.proto [hunk 2]
```diff
@@ -18,7 +18,7 @@ service Msg {
   rpc UpdateParams(MsgUpdateParams) returns (MsgUpdateParamsResponse);
 }
 
-// MsgDepositToVault is the Msg/DepositToVault request type.
+// MsgDepositToVault deposits the specified asset from the subaccout to the vault.
```
Comment (tqin7) [inline proto/dydxprotocol/vault/tx.proto:21]: nit 
```suggestion
// MsgDepositToVault deposits the specified asset from the subaccount to the vault.
```

File: proto/dydxprotocol/vault/tx.proto [hunk 3]
```diff
@@ -39,6 +39,23 @@ message MsgDepositToVault {
 // MsgDepositToVaultResponse is the Msg/DepositToVault response type.
 message MsgDepositToVaultResponse {}
 
+// MsgWithdrawFromVault withdraws the specified amount of from the vault to the subaccount.
```
Comment (tqin7) [inline proto/dydxprotocol/vault/tx.proto]: nit (depending on whether we go with shares or asset): 
```suggestion
// MsgWithdrawFromVault withdraws the specified amount of shares/asset from the vault to the subaccount.
```

File: proto/dydxprotocol/vault/tx.proto [hunk 4]
```diff
@@ -39,6 +39,23 @@ message MsgDepositToVault {
...
+  option (cosmos.msg.v1.signer) = "subaccount_id";
+  
+  // The vault to withdraw from.
+  VaultId vault_id = 1;
+  
+  // The subaccount to withdraw to.
+  dydxprotocol.subaccounts.SubaccountId subaccount_id = 2;
+  
+  // Number of shares to withdraw.
+  uint64 shares = 3;
```
Comment (tqin7) [inline proto/dydxprotocol/vault/tx.proto:53]: we use type [NumShares](https://github.com/dydxprotocol/v4-chain/blob/main/proto/dydxprotocol/vault/vault.proto#L27) to represent shares

File: proto/dydxprotocol/vault/tx.proto (hunk 1 above)
Comment (ttl33) [reply proto/dydxprotocol/vault/tx.proto]: offline conclusion: decided to go w specifying USDC

File: proto/dydxprotocol/vault/query.proto [hunk 5]
```diff
@@ -82,12 +81,6 @@ message QueryOwnerSharesRequest {
   cosmos.base.query.v1beta1.PageRequest pagination = 3;
 }
 
-// OwnerShare is a type for owner shares in a vault.
-message OwnerShare {
```
Comment (ttl33) [inline proto/dydxprotocol/vault/query.proto:86]: moved it to `vault.proto`

File: proto/dydxprotocol/vault/tx.proto [hunk 6]
```diff
@@ -39,6 +40,42 @@ message MsgDepositToVault {
...
+  // then, the amount withdrawn is $970 (= 1020 - (1000 + 50 - 1020))
+  bytes target_quote_quantums = 3 [
+    (gogoproto.customtype) =
+        "github.com/dydxprotocol/v4-chain/protocol/dtypes.SerializableInt",
+    (gogoproto.nullable) = false
+  ];
+}
+
+// MsgWithdrawFromVaultResponse is the Msg/WithdrawFromVault response type.
+message MsgWithdrawFromVaultResponse {
```
Comment (ttl33) [inline proto/dydxprotocol/vault/tx.proto:68]: note that this has been updated

File: proto/dydxprotocol/vault/tx.proto [hunk 7]
```diff
@@ -39,6 +40,42 @@ message MsgDepositToVault {
...
+  // Number of shares that have been redeemed as part of the withdrawal.
+  NumShares redeemed_shares = 1;
+
+  // Number of shares remaining after the withdrawal.
+  NumShares remaining_shares = 2;
+
+  // Number of shares for the vault after the withdrawal.
+  NumShares total_shares = 3;
```
Comment (BrendanChou) [inline proto/dydxprotocol/vault/tx.proto:76]: I assume one is the remaining shares of the individual, and one is the remaining shares in the whole vault? If so, I think this distinction could be clearer

File: proto/dydxprotocol/vault/tx.proto [hunk 8]
```diff
@@ -39,6 +48,49 @@ message MsgDepositToVault {
...
+  // The subaccount must own shares in the vault.
+  dydxprotocol.subaccounts.SubaccountId subaccount_id = 2;
+
+  // The amount of quote quantums to withdraw.
+  // The amount should account for slippage, so that the final amount withdrawn
+  // to the user matches the amount specified.
+  // If the amount specified exceeds the max amount that can be withdrawn (=
+  // user equity - slippage), only the max amount is withdrawn. For example:
+  //    * user equity = $1020
+  //    * target amount = $1000
```
Comment (tqin7) [inline proto/dydxprotocol/vault/tx.proto:72]: nit
```suggestion
  //    * quote quantums = $1000
```

File: proto/dydxprotocol/vault/tx.proto (hunk 7 above)
Comment (tqin7) [reply proto/dydxprotocol/vault/tx.proto:91]: +1. We can probably not include `total_shares` in the response?

Comment (ttl33) [reply proto/dydxprotocol/vault/tx.proto:91]: To clarify:
1. `redeemed_shares`: given that user specifies USDC, this indicates the # of shares that were redeemed for the user
2. `remaining_shares`: user's shares after the withdrawal
3. `total_shares`: shares for the entire vault

I think we need all three for the FE to calculate the new vault equity + user equity in the vault.

Technically, FE will have some of these values cached, but things could have changed in between so i think it's better to return all these so FE can calculate the amount using these values

File: proto/dydxprotocol/vault/tx.proto (hunk 4 above)
Comment (tqin7) [reply proto/dydxprotocol/vault/tx.proto]: We defined `NumShares` to 
1. use serializable int (as shares can get quite large)
2. define a message type to be used in cosmos store

Good catch in that link, it should be NumShares. I'll fix it in another PR

File: proto/dydxprotocol/vault/tx.proto (hunk 7 above)
Comment (tqin7) [reply proto/dydxprotocol/vault/tx.proto:91]: From the protocol level, does it seem a bit unexpected UX to input `quote_quantums` and in response, get `shares`?

File: proto/dydxprotocol/vault/tx.proto (hunk 1 above)
Comment (ttl33) [reply proto/dydxprotocol/vault/tx.proto]: offline conclusion: decided to go w specifying shares

File: proto/dydxprotocol/vault/tx.proto (hunk 7 above)
Comment (ttl33) [reply proto/dydxprotocol/vault/tx.proto:91]: we are using shares now per offline discussion

File: proto/dydxprotocol/vault/tx.proto [hunk 9]
```diff
@@ -39,6 +48,54 @@ message MsgDepositToVault {
...
+  NumShares redeemed_shares = 1;
+
+  // Amount of quote quantums that have been withdrawn.
+  bytes withdrawn_quote_quantums = 2 [
+    (gogoproto.customtype) =
+        "github.com/dydxprotocol/v4-chain/protocol/dtypes.SerializableInt",
+    (gogoproto.nullable) = false
+  ];
+
+  // Number of shares remaining after the withdrawal.
```
Comment (tqin7) [inline proto/dydxprotocol/vault/tx.proto:85]: nit
```suggestion
  // Number of owner shares remaining after the withdrawal.
```

File: proto/dydxprotocol/vault/tx.proto [hunk 10]
```diff
@@ -39,6 +48,54 @@ message MsgDepositToVault {
...
+  NumShares remaining_shares = 3;
+
+  // Total number of shares vault after the withdrawal.
```
Comment (tqin7) [inline proto/dydxprotocol/vault/tx.proto:88]: nit
```suggestion
  // Number of total vault shares after the withdrawal.
```

File: proto/dydxprotocol/vault/tx.proto [hunk 11]
```diff
@@ -39,6 +48,54 @@ message MsgDepositToVault {
...
+  NumShares total_vault_shares = 4;
+
+  // Total amount of quote quatums for the vault after the withdrawal.
```
Comment (tqin7) [inline proto/dydxprotocol/vault/tx.proto:91]: nit
```suggestion
  // Vault equity (in quote quantums) after withdrawal.
```

**Review by tqin7 — APPROVED** [review]

**Pull request merged**